cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -223,8 +223,13 @@ data ProjectBuildContext = ProjectBuildContext {
 runProjectPreBuildPhase
     :: Verbosity
     -> ProjectBaseContext
-    -> (ElaboratedInstallPlan -> IO ElaboratedInstallPlan)
-    -> IO ProjectBuildContext
+    -- ^ A function that selects a subset of the plan and returns it,
+    -- It can also return some arbitrary data, which can be useful in case
+    -- the caller needs it.
+    -> (ElaboratedInstallPlan -> IO (ElaboratedInstallPlan, a))
```

Comment:
Note that in that case we still have to modify all call sites to return `Nothing`. 
```haskell
... -> (ElaboratedInstallPlan -> IO (ElaboratedInstallPlan, Maybe (UnitId, ComponentId))) -> BuildCtx
```
Or use a wrapper. In the case of a monomorphic field, a *run-specific* wrapper, which I don't really like.

---

> there's only one place you need to look at to see the data that's produced in the pre-build step.

You mean the return type? I think I can agree with that, I'll add a field.

---

I'd like to see the opinion of others too on this (the polymorphism). Github suggests @dcoutts and @hvr. Even just a +1 here or there.